File: code/106.construct-binary-tree-from-inorder-and-postorder-traversal.java
/*
 * @lc app=leetcode id=106 lang=java
 *
 * [106] Construct Binary Tree from Inorder and Postorder Traversal
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, Integer> map = new HashMap<>();
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }

        return findNode(inorder, 0, inorder.length, postorder, 0, postorder.length);
    }

    public TreeNode findNode(int[] inorder, int inBegin, int inEnd, int[] postorder, int postBegin, int postEnd) {
        if (inBegin >= inEnd && postBegin >= postEnd ) {
            return null;
        }

        int rootIndex = map.get(postorder[postEnd - 1]);
        TreeNode root = new TreeNode(inorder[rootIndex]);
        int leftLength = rootIndex - inBegin;
        root.left = findNode(inorder, inBegin, rootIndex, postorder, postBegin, postBegin + leftLength);
        root.right = findNode(inorder, rootIndex + 1, inEnd, postorder , postBegin + leftLength, postEnd - 1);
        return root;
    }
}
// @lc code=end

File: code/106.construct-binary-tree-from-inorder-and-postorder-traversal.go
/*
 * @lc app=leetcode id=106 lang=golang
 *
 * [106] Construct Binary Tree from Inorder and Postorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var hashmap map[int]int

func buildTree(inorder []int, postorder []int) *TreeNode {
	hashmap = make(map[int]int)
	for i, v := range inorder {
		hashmap[v] = i
	}

	root := findNode(inorder, postorder, len(postorder)-1, 0, len(inorder)-1)
	return root
}

func findNode(inorder []int, postorder []int, rootIdx int, l, r int) *TreeNode {
	if l > r {
		return nil
	}
	if l == r {
		return &TreeNode{Val: inorder[l]}
	}

	rootV := postorder[rootIdx]
	rootIn := hashmap[rootV]
	root := &TreeNode{Val: rootV}
	root.Left = findNode(inorder, postorder, rootIdx-(r-rootIn)-1, l, rootIn-1)
	root.Right = findNode(inorder, postorder, rootIdx-1, rootIn+1, r)
	return root
}

// @lc code=end

